dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -59,23 +48,21 @@
     public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
         SettableFuture<Response> future = SettableFuture.create();
 
-        Function<Integer, ListenableFuture<Response>> callSupplier = attempt -> {
+        IntFunction<ListenableFuture<Response>> callSupplier = attempt -> {
             // TODO(dfox): include retry number in the request somehow
             return delegate.execute(endpoint, request);
         };
         FutureCallback<Response> retryer = new RetryingCallback(callSupplier, future);
-        Futures.addCallback(callSupplier.apply(0), retryer, DIRECT_EXECUTOR);
-
+        DialogueFutures.addDirectCallback(callSupplier.apply(0), retryer);
         return future;
     }
 
     private final class RetryingCallback implements FutureCallback<Response> {
         private final AtomicInteger failures = new AtomicInteger(0);
```

Comment:
This should be an `int` given that retries are sequenced we can rely on the happens-before relationship.

Reply:
do you want to make that change as part of this PR?

Reply:
I'll add it to a an improvement on top of this PR, might be best if you'd like to review that in isolation :-]